src/main/scala/io/druid/indexer/spark/SparkBatchIndexTask.scala

```diff
@@ -397,6 +401,28 @@ object SparkBatchIndexTask
         }
       )
 
+      log.info("Initializing emitter for metrics")
+
+      lifecycle.start()
+
+      closer.register(
+        new Closeable {
+          override def close(): Unit = lifecycle.stop()
+        }
+      )
+
+      sc.addSparkListener(new SparkListener() {
+        // Emit metrics at the end of each stage
+        override def onStageCompleted(event: SparkListenerStageCompleted): Unit = {
```

Comment:
Does this have a chance to fire AFTER the `closer` has fired?

In the driver the closer will fire first before the driver's method exits. If the listeners don't fire until then, there will be a nasty race on if the emitters are closed by the time they try to emit. Also, does the emitter closing attempt to deliver remaining events? if not it would be worth waiting in the listener or somewhere to make sure the events are attempted to be emitted.